### ISSUE TITLE
UDP traffic to a single Pod behind a ClusterIP is blackholed when the destination Pod is recreated (wrong conntrack)

### ISSUE THREAD

**Issue by dkulchinsky** [kubernetes/kubernetes#106274]
### What happened?

Our setup has Pods [source] that send (continuous) UDP traffic to a ClusterIP service that is backed by a single endpoint Pod [destination]

When the destination Pod is recreated, the nodes that host the source Pods end up with an incorrect conntrack entry which results in the traffic getting blackholed from the source Pod and is not delivered to the new (destination) Pod.

This seem to represent a very similar race condition as was previously found and fixed in #103857, our suspicion is the incorrect conntrack entry is created when the destination Service goes from 1 to 0 endpoints which also removes the DNAT rule, then when the destination Service goes from 0 to 1 endpoints the iptables DNAT rule is added but the conntrack has an incorrect entry from the time while traffic was still being sent from the source Pods but there was no DNAT rule in place.

---

Here's how this looks when things are broken and how we can "fix" it:

`172.18.3.25` == Source Pod IP
`172.19.3.88` == Service ClusterIP
`172.18.7.183` == Destination Pod IP (behind Service ClusterIP == `172.19.3.88`)

After the destination Pod is recreated, on the nodes that host the source Pods we see the following conntrack entry:
```
udp      17 29 src=172.18.3.25 dst=172.19.3.88 sport=48295 dport=6343 packets=403882 bytes=536947958 [UNREPLIED] src=172.19.3.88 dst=172.18.3.25 sport=6343 dport=48295 packets=0 bytes=0 mark=0 use=1
```

Note that the NATed `src` (`172.19.3.88`) here is the same as the original `dst` (`172.19.3.88`), which is incorrect and created because at this point there's no iptables DNAT rule in place (yet).

To "fix" this, we can either delete the source Pod above which forces the conntrack entry to be dropped and correctly recreated, or we can just delete the stale conntrack entry:

```
$ sudo conntrack -D -p udp --src 172.18.3.25 --dst 172.19.3.88 --sport 48295 --dport 6343
udp      17 29 src=172.18.3.25 dst=172.19.3.88 sport=48295 dport=6343 packets=363013 bytes=482736929 [UNREPLIED] src=172.19.3.88 dst=172.18.3.25 sport=6343 dport=48295 packets=0 bytes=0 mark=0 use=1
conntrack v1.4.4 (conntrack-tools): 1 flow entries have been deleted.
```

and now the conntrack is correct and traffic is flowing as expected:

```
udp      17 29 src=172.18.3.25 dst=172.19.3.88 sport=48295 dport=6343 packets=1057 bytes=1407881 [UNREPLIED] src=172.18.7.183 dst=172.18.3.25 sport=6343 dport=48295 packets=0 bytes=0 mark=0 use=1
```

/cc @aojea

### What did you expect to happen?

We expect that once the destination Pod is recreated the UDP traffic to it via the ClusterIP service will resume and work correctly.

specifically, when the destination Service goes from 0 to 1 endpoint, the iptables DNAT rule should be created (that works) and then the stale conntrack entries should be removed - we believe based on our observation that this is not taking place.

### How can we reproduce it (as minimally and precisely as possible)?

In our experience, this issue reproduces in the vast majority of cases, there were some exception which I guess depend on timing - sometimes source pods are recreated just after the destination pod as as part of a deployment so that "fixes" things, but below steps should reliably surface the issue.

1. Run a UDP Listener Pod (single pod)
2. Expose the Pod behind a ClusterIP Service
3. Send continuous stream of UDP packets from another Pod ClusterIP of the Service
4. Observe that traffic is correctly received
5. Delete the UDP Listener Pod and observe as a new one is created, traffic should now be blackholed and not reaching the "new" Pod
6. Delete the incorrect conntrack entry on the source Pod's host, traffic delivery to the destination pod will recover

### Anything else we need to know?

I did my best to see if this is an existing/known issue, and although there are various other issues around UDP/conntrack I couldn't find anything that specifically addressed the scenario I describe above.

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.2", GitCommit:"8b5a19147530eaac9476b0ab82980b4088bbc1b2", GitTreeState:"clean", BuildDate:"2021-09-15T21:31:32Z", GoVersion:"go1.16.8", Compiler:"gc", Platform:"darwin/arm64"}
Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.5-gke.1300", GitCommit:"90a16981ade07f163a0233adb631b42ac1fc53ff", GitTreeState:"clean", BuildDate:"2021-10-06T09:26:44Z", GoVersion:"go1.16.7b7", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>
GKE
</details>

### OS version

<details>

```console
$ cat /etc/os-release
NAME="Container-Optimized OS"
ID=cos
PRETTY_NAME="Container-Optimized OS from Google"
HOME_URL="https://cloud.google.com/container-optimized-os/docs"
BUG_REPORT_URL="https://cloud.google.com/container-optimized-os/docs/resources/support-policy#contact_us"
GOOGLE_METRICS_PRODUCT_ID=26
KERNEL_COMMIT_ID=6a122d2c907a3fe5fa9215ce7a5c989a17f785c2
GOOGLE_CRASH_ID=Lakitu
VERSION=89
VERSION_ID=89
BUILD_ID=16108.534.8

$ uname -a
Linux gke-prd-usc1-workloads-4fae8c9b-z25p 5.4.144+ #1 SMP Sat Sep 25 09:56:01 PDT 2021 x86_64 Intel(R) Xeon(R) CPU @ 2.80GHz GenuineIntel GNU/Linux
```

</details>

### Install tools

<details>
GKE
</details>

### Container runtime (CRI) and and version (if applicable)

<details>
docker://20.10.3
</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>
Calico
</details>

**Comment by aojea** [comment]
We have an e2e test with the same scenario that you are describing, however, we recently had a regression with unready pods 
https://github.com/kubernetes/kubernetes/pull/106163

However, I assume the conntrack entry is obtained from the node

> udp      17 29 src=172.18.3.25 dst=172.19.3.88 sport=48295 dport=6343 packets=403882 bytes=536947958 [UNREPLIED] src=172.19.3.88 dst=172.18.3.25 sport=6343 dport=48295 packets=0 bytes=0 mark=0 use=1

> 172.18.3.25 == Source Pod IP

is the source the pod a host network pod? i.e. 172.18.3.25 is also the IP of the node?
otherwise I can't understand that conntrack entry ... :thinking:

**Comment by dkulchinsky** [comment]
Hi @aojea! thanks for taking a look.

both Pods (source & destination) are not on `hostNetwork`, our Pods CIDR for this cluster is: `172.18.0.0/16`

These are the pods in question (names obfuscated):
```
<source pod>-7ff6c6755c-7l9q5              2/2     Running   0          4d8h    172.18.3.25    gke-<cluster>-workloads-4fae8c9b-z25p   <none>           <none>
<source pod>-7ff6c6755c-bqwhs              2/2     Running   0          4d7h    172.18.1.8     gke-<cluster>-workloads-0d3fd2ae-y9l9   <none>           <none>
<destination pod>-f56bfcbf4-vrq9w                  2/2     Running   0          9h      172.18.7.183   gke-<cluster>-workloads-0d3fd2ae-tcwd   <none>           <none>
```

**Comment by aojea** [comment]
let me know if you can dump the iptables rules at the moment of that entry get created ... and the yaml for the pods and the service (is ok to obfuscate the name or fields you consider )

**Comment by dkulchinsky** [comment]
@aojea I happened to save these 😄 

`iptables.txt` includes the iptables when traffic from Pod with IP 172.18.3.25 was being blackholed.

[iptables.txt](https://github.com/kubernetes/kubernetes/files/7508983/iptables.txt)

here are the manifests for the Pods and Service (sorry, I had to obfuscate a bunch of details).

Source Pod:

<details>

```yaml
apiVersion: v1
kind: Pod
metadata:
  annotations:
    cni.projectcalico.org/containerID: ee467bbbc59841b4c8b1bcc7c4dfce234b116e639ef6b121d59b574479a0e9f8
    cni.projectcalico.org/podIP: 172.18.3.25/32
    cni.projectcalico.org/podIPs: 172.18.3.25/32
    kubernetes.io/psp: mesh-privileged
    policies.kyverno.io/last-applied-patches: |
      add-imagepullsecrets.add-harbor-image-pull-secret.kyverno.io: added /spec/imagePullSecrets/0
    prometheus.io/path: /metrics
    prometheus.io/port: "65100"
    prometheus.io/scrape: "true"
  labels:
    app.kubernetes.io/component: <src>
    app.kubernetes.io/name: <src>
    pod-template-hash: 7ff6c6755c
  name: <src>-7ff6c6755c-7l9q5
  ownerReferences:
  - apiVersion: apps/v1
    blockOwnerDeletion: true
    controller: true
    kind: ReplicaSet
    name: <src>-7ff6c6755c
    uid: 85195b28-d7bc-4b8f-9b11-ef83ac73dd0b
  resourceVersion: "224306681"
  uid: 79db14bd-d8be-4d71-8ac8-e7590809934f
spec:
  affinity:
    podAntiAffinity:
      preferredDuringSchedulingIgnoredDuringExecution:
      - podAffinityTerm:
          labelSelector:
            matchExpressions:
            - key: app.kubernetes.io/component
              operator: In
              values:
              - <src>
          topologyKey: kubernetes.io/hostname
        weight: 100
  containers:
  - args:
    - --mode
    - server
    - --listen
    - :4433
    - --backend
    -<dst service>.svc:6343
    image: <registry>/<src image>:0c5034a56d30-62
    imagePullPolicy: IfNotPresent
    livenessProbe:
      failureThreshold: 3
      httpGet:
        path: /health
        port: 65100
        scheme: HTTP
      initialDelaySeconds: 5
      periodSeconds: 5
      successThreshold: 1
      timeoutSeconds: 1
    name: <src>
    ports:
    - containerPort: 4433
      protocol: UDP
    - containerPort: 65100
      protocol: TCP
    resources:
      limits:
        cpu: "4"
        memory: 1Gi
      requests:
        cpu: 100m
        memory: 250Mi
    securityContext:
      allowPrivilegeEscalation: false
      capabilities:
        drop:
        - all
      privileged: false
      readOnlyRootFilesystem: true
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-dq5x8
      readOnly: true
  dnsPolicy: ClusterFirst
  enableServiceLinks: true
  imagePullSecrets:
  - name: harbor
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: FallbackToLogsOnError
    volumeMounts:
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-dq5x8
      readOnly: true
  nodeName: gke-prd-usc1-workloads-4fae8c9b-z25p
  preemptionPolicy: PreemptLowerPriority
  priority: 0
  restartPolicy: Always
  schedulerName: default-scheduler
  securityContext:
    fsGroup: 10001
    runAsGroup: 10001
    runAsNonRoot: true
    runAsUser: 10001
  serviceAccount: telemetry-pmacct
  serviceAccountName: telemetry-pmacct
  terminationGracePeriodSeconds: 30
  tolerations:
  - effect: NoExecute
    key: node.kubernetes.io/not-ready
    operator: Exists
    tolerationSeconds: 300
  - effect: NoExecute
    key: node.kubernetes.io/unreachable
    operator: Exists
    tolerationSeconds: 300
  volumes:
  - name: kube-api-access-dq5x8
    projected:
      defaultMode: 420
      sources:
      - serviceAccountToken:
          expirationSeconds: 3607
          path: token
      - configMap:
          items:
          - key: ca.crt
            path: ca.crt
          name: kube-root-ca.crt
      - downwardAPI:
          items:
          - fieldRef:
              apiVersion: v1
              fieldPath: metadata.namespace
            path: namespace
```

</details>


Destination Pod:

<details>

```yaml
apiVersion: v1
kind: Pod
metadata:
  annotations:
    cni.projectcalico.org/containerID: 30c97c318ad004f3f47a43bf3e11236dda04cf24870b324b0abf0dc787953241
    cni.projectcalico.org/podIP: 172.18.7.183/32
    cni.projectcalico.org/podIPs: 172.18.7.183/32
    kubernetes.io/psp: mesh-privileged
    policies.kyverno.io/last-applied-patches: |
      add-imagepullsecrets.add-harbor-image-pull-secret.kyverno.io: added /spec/imagePullSecrets/0
  labels:
    app.kubernetes.io/component: <dst>
    app.kubernetes.io/name: <dst>
    pod-template-hash: f56bfcbf4
  name: <dst>-f56bfcbf4-vrq9w
  ownerReferences:
  - apiVersion: apps/v1
    blockOwnerDeletion: true
    controller: true
    kind: ReplicaSet
    name: <dst>-f56bfcbf4
    uid: f2993c4b-6307-4886-9132-8dd6933eb7bb
  resourceVersion: "229519058"
  uid: 67246a00-38ad-4f66-a879-0d6b0d41da5f
spec:
  affinity:
    podAntiAffinity:
      preferredDuringSchedulingIgnoredDuringExecution:
      - podAffinityTerm:
          labelSelector:
            matchExpressions:
            - key: app.kubernetes.io/component
              operator: In
              values:
              - <dst>
          topologyKey: kubernetes.io/hostname
        weight: 100
  containers:
  - env:
    - name: KAFKA_KEY_PASSWORD
      valueFrom:
        secretKeyRef:
          key: user.password
          name: <dst>
    envFrom:
    - configMapRef:
        name: <dst>-env-vars
    image: <registry>/<dst-image>:5d46efcc29f0-78
    imagePullPolicy: IfNotPresent
    livenessProbe:
      exec:
        command:
        - /bin/bash
        - -c
        - if test $(pidof -o %PPID <dst> | wc -w) -lt 2 ; then exit 1; else exit
          0; fi
      failureThreshold: 3
      initialDelaySeconds: 5
      periodSeconds: 5
      successThreshold: 1
      timeoutSeconds: 1
    name: <dst>
    ports:
    - containerPort: 6343
      protocol: UDP
    - containerPort: 17917
      protocol: TCP
    resources:
      limits:
        cpu: "16"
        memory: 2Gi
      requests:
        cpu: "1"
        memory: 650Mi
    securityContext:
      allowPrivilegeEscalation: false
      capabilities:
        drop:
        - all
      privileged: false
      readOnlyRootFilesystem: true
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /etc/<path>
      name: kafka-autoprovisioned-secrets
      readOnly: true
    - mountPath: /etc/pmacct
      name: etc-pmacct
    - mountPath: /tmp
      name: tmp
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-8drmf
      readOnly: true
  dnsPolicy: ClusterFirst
  enableServiceLinks: true
  imagePullSecrets:
  - name: harbor
  nodeName: gke-prd-usc1-workloads-0d3fd2ae-tcwd
  preemptionPolicy: PreemptLowerPriority
  priority: 0
  restartPolicy: Always
  schedulerName: default-scheduler
  securityContext:
    fsGroup: 10001
    runAsGroup: 10001
    runAsNonRoot: true
    runAsUser: 10001
  serviceAccount: telemetry-pmacct
  serviceAccountName: telemetry-pmacct
  terminationGracePeriodSeconds: 30
  tolerations:
  - effect: NoExecute
    key: node.kubernetes.io/not-ready
    operator: Exists
    tolerationSeconds: 300
  - effect: NoExecute
    key: node.kubernetes.io/unreachable
    operator: Exists
    tolerationSeconds: 300
  volumes:
  - name: kafka-autoprovisioned-secrets
    secret:
      defaultMode: 420
      items:
      - key: ca.crt
        mode: 420
        path: ca.crt
      - key: user.key
        mode: 416
        path: user.key
      - key: user.crt
        mode: 420
        path: user.crt
      secretName: <dst>
  - emptyDir: {}
    name: etc-pmacct
  - emptyDir: {}
    name: tmp
  - name: kube-api-access-8drmf
    projected:
      defaultMode: 420
      sources:
      - serviceAccountToken:
          expirationSeconds: 3607
          path: token
      - configMap:
          items:
          - key: ca.crt
            path: ca.crt
          name: kube-root-ca.crt
      - downwardAPI:
          items:
          - fieldRef:
              apiVersion: v1
              fieldPath: metadata.namespace
            path: namespace
```

</details>


and the Service:

<details>

```yaml

apiVersion: v1
kind: Service
metadata:
  name: <dst service>
spec:
  clusterIP: 172.19.3.88
  clusterIPs:
  - 172.19.3.88
  ipFamilies:
  - IPv4
  ipFamilyPolicy: SingleStack
  ports:
  - name: sflow
    port: 6343
    protocol: UDP
    targetPort: 6343
  selector:
    app.kubernetes.io/component: <dst>
    app.kubernetes.io/name: <dst>
  sessionAffinity: None
  type: ClusterIP
status:
  loadBalancer: {}

```

</details>

**Comment by aojea** [comment]
hmm
the iptables rules pasted show that the service has active endpoints

> -A KUBE-SERVICES ! -s 172.18.3.0/24 -d 172.19.3.88/32 -p udp -m comment --comment "autopilot/sfacctd-sjc100-sflow-service:sflow cluster IP" -m udp --dport 6343 -j KUBE-MARK-MASQ
> -A KUBE-SERVICES -d 172.19.3.88/32 -p udp -m comment --comment "autopilot/sfacctd-sjc100-sflow-service:sflow cluster IP" -m udp --dport 6343 -j KUBE-SVC-LMENNNDADLUFQBWB

> -A KUBE-SVC-LMENNNDADLUFQBWB -m comment --comment "autopilot/sfacctd-sjc100-sflow-service:sflow" -j KUBE-SEP-FI4JKBA7XKLSZU7O

> -A KUBE-SEP-FI4JKBA7XKLSZU7O -s 172.18.7.183/32 -m comment --comment "autopilot/sfacctd-sjc100-sflow-service:sflow" -j KUBE-MARK-MASQ
> -A KUBE-SEP-FI4JKBA7XKLSZU7O -p udp -m comment --comment "autopilot/sfacctd-sjc100-sflow-service:sflow" -m udp -j DNAT --to-destination 172.18.7.183:6343

if you say that the pod was restarting this may be caused by the regression fixed in 

I still don't get why the pod traffic is being hairpined to the original pod, it seems that the source pod belong to another service too ...

It will be nice to discard that your are not impacted by these 2 known issues with unready endpoints, maybe kube-proxy logs during the occurrence can show more light

https://github.com/kubernetes/kubernetes/pull/106163
https://github.com/kubernetes/kubernetes/pull/106030

**Comment by dkulchinsky** [comment]
> hmm the iptables rules pasted show that the service has active endpoints
> 
> > -A KUBE-SERVICES ! -s 172.18.3.0/24 -d 172.19.3.88/32 -p udp -m comment --comment "autopilot/sfacctd-sjc100-sflow-service:sflow cluster IP" -m udp --dport 6343 -j KUBE-MARK-MASQ
> > -A KUBE-SERVICES -d 172.19.3.88/32 -p udp -m comment --comment "autopilot/sfacctd-sjc100-sflow-service:sflow cluster IP" -m udp --dport 6343 -j KUBE-SVC-LMENNNDADLUFQBWB
> 
> > -A KUBE-SVC-LMENNNDADLUFQBWB -m comment --comment "autopilot/sfacctd-sjc100-sflow-service:sflow" -j KUBE-SEP-FI4JKBA7XKLSZU7O
> 
> > -A KUBE-SEP-FI4JKBA7XKLSZU7O -s 172.18.7.183/32 -m comment --comment "autopilot/sfacctd-sjc100-sflow-service:sflow" -j KUBE-MARK-MASQ
> > -A KUBE-SEP-FI4JKBA7XKLSZU7O -p udp -m comment --comment "autopilot/sfacctd-sjc100-sflow-service:sflow" -m udp -j DNAT --to-destination 172.18.7.183:6343
> 
> if you say that the pod was restarting this may be caused by the regression fixed in

The Destination pod (sfacctd) was recreated as part of a deployment update (new image tag)

I assume you refer to #106163 here?

> I still don't get why the pod traffic is being hairpined to the original pod, 

mmm, not the original Pod but the Service Cluster IP, right?

this:
```
... [UNREPLIED] src=172.19.3.88 dst=172.18.3.25 sport=6343 dport=48295
```

should be this:
```
... [UNREPLIED] src=172.18.7.183 dst=172.18.3.25 sport=6343 dport=48295
```

where `172.18.7.183` is the sfacctd Pod IP (the destination).

> it seems that the source pod belong to another service too ...

Correct, the flow-proxy Pods are behind a LoadBalancer Service and receives traffic from external sources, this is where we hit #103857 which is now fixed thanks to you  😄 

our full flow looks like this:

```
[external sources]  --> [LoadBalancer Service] --> [flow-proxy] --> [ClusterIP Service] --> [sfacctd]
```

> It will be nice to discard that your are not impacted by these 2 known issues with unready endpoints, maybe kube-proxy logs during the occurrence can show more light

I combed through the kube-proxy logs (log level is set to `--v=2`) on the node where the "broken" source Pod was running at the time when the destination Pod was being recreated and didn't see anything of interest, mostly:
- `Syncing iptables rules`
- `syncProxyRules complete`

Is there a specific type of message you were expecting to see? maybe I messed up my logs search.

> 
> #106163 #106030

I'd love to test these out, but since we use GKE, I'm not sure there's a way to go about it until a release with these patches is available in GKE.

**Comment by dkulchinsky** [comment]
Hey @aojea, any thoughts or updates about this?

**Comment by aojea** [comment]
> Hey @aojea, any thoughts or updates about this?
> ```
> [external sources]  --> [LoadBalancer Service] --> [service A (source)] --> [ClusterIP Service] --> [service B (destination)]
> ```

I'm really busy these days, but my suggestion is to try to check which parts of ^^^ this path ^^^ works and which parts don't , i.e you can try to see if it ONLY happens for traffic that comes through the load balancer, and not if is generated within the cluster or ....

**Comment by dkulchinsky** [comment]
> > Hey @aojea, any thoughts or updates about this?
> > ```
> > [external sources]  --> [LoadBalancer Service] --> [service A (source)] --> [ClusterIP Service] --> [service B (destination)]
> > ```
> 
> I'm really busy these days

Understood and thanks for taking the time 👍🏼 
but my suggestion is to try to check which parts of ^^^ this path ^^^ works and which parts don't , i.e you can try to see if it ONLY happens for traffic that comes through the load balancer, and not if is generated within the cluster or ....

**Comment by dkulchinsky** [comment]
> > Hey @aojea, any thoughts or updates about this?
> > ```
> > [external sources]  --> [LoadBalancer Service] --> [service A (source)] --> [ClusterIP Service] --> [service B (destination)]
> > ```
> 
> I'm really busy these days, but my suggestion is to try to check which parts of ^^^ this path ^^^ works and which parts don't

Thanks for taking the time to look into this @aojea!

(apologies for closing/reopening... finger slipped, haha)


This path works as expected:
```
[external sources]  --> [LoadBalancer Service] --> [service A (source)] 
```

This is where we see the issue:
```
[service A (source)] --> [ClusterIP Service] --> [service B (destination)]
```

>  i.e you can try to see if it ONLY happens for traffic that comes through the load balancer, and not if is generated within the cluster or ....

The issue occurs only for traffic originated from Pod(s) running in the cluster (`[service A]` above), the fact that `[service A]` also receives packets from external sources shouldn't be a factor here, isn't it?

The packets from external sources are terminated by `[service A]` (these are DTLS packets), they are decrypted and processed, then a new socket/connection is created from `[service A]` to `[service B]` via a `[ClusterIP Service]` that sends that data in unencrypted UDP packets.

in every case that we've examined where this issue happened, the `conntrack` entry was incorrect and was fixed by simply deleting the bad conntrack entry and a new/correct one is created and everything starts working as expected.

in the meantime, we've made the following change to workaround this issue: since `[service B]` is always backed by a single Pod, we changed the `[ClusterIP Service]` that fronts it to be a `headless` service, we have a mechanism in `[service A]` to re-check DNS every one minute and the application sends the traffic directly to `[service B]` Pod IP.

The above works well and satisfies our needs at the moment, there are plans to allow multiple replicas of `[service B]` so we will need to revisit this approach then, but perhaps we won't be hitting this issue when `[service B]` will always have >1 replicas (?)
